src/Hosting/Hosting/src/Internal/HostingRequestFinishedLog.cs

```diff
@@ -27,33 +29,49 @@ internal class HostingRequestFinishedLog : IReadOnlyList<KeyValuePair<string, ob
-                        return new KeyValuePair<string, object>("ElapsedMilliseconds", _elapsed.TotalMilliseconds);
+                        return new KeyValuePair<string, object>("ElapsedMilliseconds", Elapsed.TotalMilliseconds);
                     case 1:
-                        return new KeyValuePair<string, object>("StatusCode", _httpContext.Response.StatusCode);
+                        return new KeyValuePair<string, object>("StatusCode", StatusCode);
                     case 2:
-                        return new KeyValuePair<string, object>("ContentType", _httpContext.Response.ContentType);
+                        return new KeyValuePair<string, object>("ContentType", ContentType);
+                    case 3:
+                        return new KeyValuePair<string, object>("ContentLength", ContentLength);
+                    case 4:
+                        return new KeyValuePair<string, object>("Protocol", _startLog.Protocol);
+                    case 5:
+                        return new KeyValuePair<string, object>("Method", _startLog.Method);
+                    case 6:
+                        return new KeyValuePair<string, object>("Scheme", _startLog.Scheme);
+                    case 7:
+                        return new KeyValuePair<string, object>("Host", _startLog.Host);
+                    case 8:
+                        return new KeyValuePair<string, object>("PathBase", _startLog.PathBase);
+                    case 9:
+                        return new KeyValuePair<string, object>("Path", _startLog.Path);
+                    case 10:
+                        return new KeyValuePair<string, object>("QueryString", _startLog.QueryString);
                     default:
                         throw new IndexOutOfRangeException(nameof(index));
                 }
             }
         }
 
-        public HostingRequestFinishedLog(HttpContext httpContext, TimeSpan elapsed)
+        public HostingRequestFinishedLog(HostingRequestStartingLog startLog, int statusCode, long? contentLength, string contentType, TimeSpan elapsed)
         {
-            _httpContext = httpContext;
-            _elapsed = elapsed;
+            _startLog = startLog;
+
+            ContentType = !string.IsNullOrEmpty(contentType) ? Uri.EscapeUriString(contentType) : string.Empty;
+            ContentLength = contentLength;
+            StatusCode = statusCode;
+            Elapsed = elapsed;
         }
 
         public override string ToString()
         {
             if (_cachedToString == null)
             {
-                _cachedToString = string.Format(
-                    CultureInfo.InvariantCulture,
-                    "Request finished in {0}ms {1} {2}",
-                    _elapsed.TotalMilliseconds,
-                    _httpContext.Response.StatusCode,
-                    _httpContext.Response.ContentType);
+                var logStart = _startLog.ToString().Substring(HostingRequestStartingLog.LogPreamble.Length);
```

Comment:
What's the cause with all the space before the dash in the "After:" sample output? Is HostingRequestStartingLog.ToString() returning a lot of trailing spaces?

```
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/1.1 GET http://ben-laptop/
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished HTTP/1.1 GET http://ben-laptop/   - 404 0  27.9983ms
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/1.1 GET http://ben-laptop/index.html
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished HTTP/1.1 GET http://ben-laptop/index.html   - 200 2348 text/html 51.6009ms
i
```

Reply:
Space separated fields and they are empty.

Is why I `Uri.EscapeUriString` on content type so `application/json; charset=utf-8` would come out `application/json;+charset=utf-8` as IIS does:

![image](https://user-images.githubusercontent.com/1142958/57496568-a691a080-72cb-11e9-8317-7d14919d809b.png)

Reply:
Should we trim the string then?

```suggestion
                var logStart = _startLog.ToString().Substring(HostingRequestStartingLog.LogPreamble.Length).TrimEnd(' ');
```

Reply:
Or better yet, have HostingRequestStartingLog.ToString() do that.

Reply:
If you did a `string.Split(` `)` on it wouldn't you want the same number of fields in the same positions?

Reply:
I'm not suggesting `string.Split(' ')`, but instead `string.TrimEnd(' ')`.

Reply:
I mean as the consumer of the logs; if the end (in this case the middle) was trimmed then space would no longer be a delimiter as the value in each column would be determined by the preceding fields and not by its position.

e.g. if this were instead csv it would look like

```
Request,finished,HTTP/1.1,GET,http://ben-laptop/textures/spacedrop2048_back6.dds,,,-,200,1445726,application/octet-stream,672.4511ms
```
Collating the commas would be problematic as the values would keep changing columns?

Reply:
e.g. a Post request would look like
```
Request finished HTTP/1.1 POST http://ben-laptop/myform application/json;+charset=utf-8 2567 - 200 1445726 application/octet-stream 672.4511ms
```
Or with commas
```
Request,finished,HTTP/1.1,POST,http://ben-laptop/myform,application/json;+charset=utf-8,2567,-,200,1445726,application/octet-stream,672.4511ms
```

Reply:
I hold the principle that object.ToString() (in this case HostingRequestStartingLog.ToString()) should *never* have trailing spaces.

Whatever is interpolating object.ToString() into anther string (e.g. string.Format) should add any necessary surrounding whitespace.

Reply:
I do see the point about parsing logs though. If we used commas instead of spaces at least it would be clear that there are empty columns. Now anyone reading the logs just sees what looks like bad formatting.

Reply:
But anyone parsing the logs should be using semantic logged output, right?

Reply:
> But anyone parsing the logs should be using semantic logged output, right?

What people do, and what they should do, are often very far apart 😉 

These issues would suggest not always https://github.com/aspnet/AspNetCore/issues/5894, https://github.com/aspnet/AspNetCore/issues/10097

Reply:
Also I do know logging at Info in production is far more common than I thought it was,

Reply:
Can we use something other than a space as the column separator? Maybe '|' or ','? We'd have to make sure whatever separator we chose keeps the logs easily readable, doesn't just look like bad formatting, and doesn't conflict with characters that could show up inside a column entry.

For a second I thought a ' ' in the ContentType column might ruin using ' ' as a column separator, but the Content-Type header gets url-encoded.

Reply:
Was looking at the w3c format and it goes with
> If a field is unused in a particular entry dash "-" marks the omitted field.

Might be better?

Reply:
Yes! That's much better IMHO.

Reply:
Now outputs:
```
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/1.1 GET http://ben-laptop/index.html - -
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished HTTP/1.1 GET http://ben-laptop/index.html - - - 200 2348 text/html 76.8050ms
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/1.1 GET http://ben-laptop/js/aoa.avatar.js - -
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished HTTP/1.1 GET http://ben-laptop/js/aoa.avatar.js - - - 200 39852 application/octet-stream 8.6488ms
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/1.1 POST http://ben-laptop/form application/x-www-form-urlencoded 19
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished HTTP/1.1 POST http://ben-laptop/form application/x-www-form-urlencoded 19 - 404 0 - 8.1129ms
```